### PR TITLE
intel_adsp: ace: secondary core power-up

### LINKED ISSUE #54327: intel_adsp: ace: various multicore bugs, timeouts
**Describe the bug**

- Core1 and Core2 is still enabled after putting into D3
- The core does not go into a lower power state in the expected time because the idle thread tick is too big.
- Timeout on reenabling core 1 

**To Reproduce**

This can be reproduced when testing Zephyr with SOF.

**Expected behavior**
No timeouts multicore operating as expected 

**Impact**

Communication timeouts with the host when using SOF.

### PR DESCRIPTION
**Pull request by tmleman**
Changes in the sequence of entering and exiting D3 state by the secondary cores.

### CODE REVIEW

File: soc/xtensa/intel_adsp/ace/power.c [hunk 1]
```diff
@@ -220,7 +220,9 @@ __weak void pm_state_set(enum pm_state state, uint8_t substate_id)
 		z_xt_ints_off(0xffffffff);
 		core_desc[cpu].bctl = DSPCS.bootctl[cpu].bctl;
 		DSPCS.bootctl[cpu].wdtcs = DSPBR_WDT_RESTART_COMMAND;
+		DSPCS.bootctl[cpu].bctl &= ~DSPBR_BCTL_WAITIPPG;
 		DSPCS.bootctl[cpu].bctl &= ~DSPBR_BCTL_WAITIPCG;
```
Comment (softwarecki) [inline soc/xtensa/intel_adsp/ace/power.c:224]: DSPCS.bootctl[cpu].bctl &= ~(DSPBR_BCTL_WAITIPPG | DSPBR_BCTL_WAITIPCG);

File: soc/xtensa/intel_adsp/ace/multiprocessing.c [hunk 2]
```diff
@@ -96,6 +96,15 @@ void soc_start_core(int cpu_num)
 		DSPCS.bootctl[cpu_num].battr |= DSPBR_BATTR_LPSCTL_BATTR_SLAVE_CORE;
 	}
 
+	/* Disabling power before core power up. This step is required if core is starting after
+	 * entering power gating and being turn off.
+	 */
```
Comment (softwarecki) [inline soc/xtensa/intel_adsp/ace/multiprocessing.c:101]: shouldn't there be a "Disabling power **gating** before ..."

Comment (ceolin) [reply soc/xtensa/intel_adsp/ace/multiprocessing.c:101]: yep, this comment is a little bit hard to understand

File: soc/xtensa/intel_adsp/ace/multiprocessing.c [hunk 3]
```diff
@@ -96,6 +96,15 @@ void soc_start_core(int cpu_num)
...
+	DSPCS.capctl[cpu_num].ctl &= ~DSPCS_CTL_SPA;
+
+	while (((DSPCS.capctl[cpu_num].ctl & DSPCS_CTL_CPA) == DSPCS_CTL_CPA)) {
```
Comment (ceolin) [inline soc/xtensa/intel_adsp/ace/multiprocessing.c:104]: It is worth some comment here about what you are checking. These register names are not easy to udnerstand

**Comment by ceolin** [comment]
The logic in this pr looks fine for me, please take a look in the comments

**Comment by marcinszkudlinski** [comment]
looks fine

File: soc/xtensa/intel_adsp/ace/multiprocessing.c (hunk 2 above)
Comment (tmleman) [reply soc/xtensa/intel_adsp/ace/multiprocessing.c:102]: This bit is not responsible for power gating. The comment has been updated.

File: soc/xtensa/intel_adsp/ace/multiprocessing.c [hunk 4]
```diff
@@ -170,6 +181,7 @@ int soc_adsp_halt_cpu(int id)
 		return -EINVAL;
 	}
 
+	ACE_PWRCTL->wpdsphpxpg &= ~BIT(id);
```
Comment (softwarecki) [inline soc/xtensa/intel_adsp/ace/multiprocessing.c:184]: What is wpdsphpxpg?

Comment (tmleman) [reply soc/xtensa/intel_adsp/ace/multiprocessing.c:184]: Commit message is describing what this bitfield does. In the future, it should be done with usage of power_domain_intel_adsp.

**Review by aborisovich — APPROVED** [review]
Looks good to me

**Review by dcpleung — APPROVED** [review]

File: soc/xtensa/intel_adsp/ace/multiprocessing.c (hunk 4 above)
Comment (softwarecki) [reply soc/xtensa/intel_adsp/ace/multiprocessing.c:184]: When you read a file and try to understand what it does, you don't do git blame to read all the commit messages. The power control code is full of names that look like the result of obfuscation (eg. wpdsphpxpg, wphstpg, wphubhppg, wpdspulppg, wpioxpg, wpmlpg, phubulppg, dsphpxpgs, hstpgs, hubhppgs, dspulppgs, ioxpgs, mlpgs, hubulppgs). We should definitely add comments with human names, add one-sentence descriptions of what we're doing when toggling some bit. Without it, it's very easy to make a mistake and code maintenance becomes impossible.

**Review by marcinszkudlinski — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Comment by stephanosio** [comment]
@tmleman A bug report is required in order for this PR to be merged during the feature freeze. Please file a bug report issue for the problem that this PR is trying to fix and link the issue to this PR.

**Comment by nashif** [comment]
@stephanosio  bug added

**Pull request merged**